storage: transformers: pass a context.Context

stevekuznetsov opened the pull request:
When an envelope transformer calls out to KMS (for instance), it will be
very helpful to pass a `context.Context` to allow for cancellation. This
patch does that, while passing the previously-expected additional data
via a context value.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/kind cleanup
/sig api-machinery

```release-note
NONE
```

```docs

```

/cc @deads2k @liggitt @enj

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -199,13 +200,13 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
 	}
 
 	for _, testCase := range transformers {
-		transformedData, err := testCase.Transformer.TransformToStorage(originalText, context)
```
The diff is a little larger than it has to be since I renamed these so as to not collide with the package name.

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -2250,7 +2250,7 @@ func TestConsistentList(t *testing.T) {
 
 	for i := 0; i < 5; i++ {
-		if err := transformer.createObject(); err != nil {
+		if err := transformer.createObject(context.TODO()); err != nil {
```
My follow-up will be to change these, at which point only the top-level contexts will need any change.

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:
```diff
@@ -45,20 +64,20 @@ type Transformer interface {
 	// Stale is true if the object on disk is stale and a write to etcd should be issued, even if the contents of the object
 	// have not changed.
-	TransformFromStorage(data []byte, context Context) (out []byte, stale bool, err error)
+	TransformFromStorage(data []byte, context context.Context) (out []byte, stale bool, err error)
```
I expected `TransformFromStorage(ctx context.Context, data []byte, dataContext Context)`, not plumbing the existing second parameter as a `context.Context` value. The colliding interface name is unfortunate, but if we were going to fix that, a rename of the param and/or the interface could work

stevekuznetsov replied:
Yep, I realized after doing this that you wanted that change. Will do :)

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:
```diff
@@ -45,20 +64,20 @@ type Transformer interface {
+	TransformFromStorage(data []byte, context context.Context) (out []byte, stale bool, err error)
 	// TransformToStorage may transform the provided data into the appropriate form in storage or return an error.
-	TransformToStorage(data []byte, context Context) (out []byte, err error)
+	TransformToStorage(data []byte, context context.Context) (out []byte, err error)
```
same here (keeping the existing explicit typed parameter, and adding `ctx context.Context` as the first param)

stevekuznetsov committed:
storage: transformers: pass a context.Context

When an envelope transformer calls out to KMS (for instance), it will be
very helpful to pass a `context.Context` to allow for cancellation. This
patch does that, while passing the previously-expected additional data
via a context value.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -79,24 +79,24 @@ type prefixTransformer struct {
-func (p *prefixTransformer) TransformFromStorage(b []byte, ctx value.Context) ([]byte, bool, error) {
+func (p *prefixTransformer) TransformFromStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, bool, error) {
 	atomic.AddUint64(&p.reads, 1)
 	if ctx == nil {
```
```suggestion
	if dataCtx == nil {
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_test.go:
```diff
@@ -78,7 +80,7 @@ func TestGCMKeyRotation(t *testing.T) {
 
 	// verify changing the context fails storage
-	_, _, err = p.TransformFromStorage(out, value.DefaultContext([]byte("incorrect_context")))
+	_, _, err = p.TransformFromStorage(context.Background(), out, value.DefaultContext([]byte("incorrect_context")))
```
```suggestion
	_, _, err = p.TransformFromStorage(ctx, out, value.DefaultContext([]byte("incorrect_context")))
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope_test.go:
```diff
@@ -230,14 +234,15 @@ func TestBackwardsCompatibility(t *testing.T) {
 	originalText := []byte(testText)
 
-	transformedData, err := oldTransformToStorage(envelopeTransformerInst.(*envelopeTransformer), originalText, context)
+	transformedData, err := oldTransformToStorage(envelopeTransformerInst.(*envelopeTransformer), ctx, originalText, dataCtx)
```
nit: make ctx the first param

stevekuznetsov replied:
ACK, was trying to keep the (receiver, args) pattern

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/secretbox/secretbox_test.go:
```diff
@@ -58,7 +60,7 @@ func TestSecretboxKeyRotation(t *testing.T) {
 	// verify changing the context does not fails storage
 	// Secretbox is not currently an authenticating store
-	_, _, err = p.TransformFromStorage(out, value.DefaultContext([]byte("incorrect_context")))
+	_, _, err = p.TransformFromStorage(context.Background(), out, value.DefaultContext([]byte("incorrect_context")))
```
```suggestion
	_, _, err = p.TransformFromStorage(ctx, out, value.DefaultContext([]byte("incorrect_context")))
```

liggitt reviewed (commented):
a few tweaks on tests, lgtm otherwise

liggitt commented:
/lgtm
/approve

fedebongio commented:
/triage accepted
Thank you Jordan for reviewing the entire series! @liggitt
